app/uk/gov/hmrc/agentsubscriptionfrontend/controllers/SubscriptionController.scala

```diff
@@ -187,33 +208,81 @@ class SubscriptionController @Inject()(
+        }
+      }
+    }
+  }
+
+  val submitLinkAccount: Action[AnyContent] = Action.async { implicit request =>
+    withAuthenticatedAgent {
+      request.session.get("arn").fold(Future.successful(sessionMissingRedirect())) { arn =>
+        linkAccountForm
+          .bindFromRequest()
+          .fold(
+            formWithErrors => {
+              if (formWithErrors.errors.exists(_.message == "error.link-account-value.invalid")) {
+                throw new BadRequestException("Form submitted with strange input value")
+              } else {
+                Future.successful(Ok(html.link_account(formWithErrors)))
+              }
+            },
+            validatedLinkAccount => {
+              validatedLinkAccount.autoMapping match {
+                case Yes => linkAccountResponse(mappingConnector.updatePreSubscriptionWithArn)
+                case No  => linkAccountResponse(mappingConnector.deletePreSubscription)
+              }
+            }
+          )
+      }
+    }
+  }
+
   val showSubscriptionComplete: Action[AnyContent] = Action.async { implicit request =>
+    def recoverSessionStoreWithNone[T]: PartialFunction[Throwable, Option[T]] = {
+      case NonFatal(ex) =>
+        Logger(getClass).warn("Session store service failure", ex)
+        None
+    }
+
     withAuthenticatedAgent {
-      request.flash.get("arn") match {
+      request.session.get("arn") match {
         case Some(arn) =>
-          sessionStoreService.fetchContinueUrl
-            .recover {
-              case NonFatal(ex) =>
-                Logger(getClass).warn("Session store service failure", ex)
-                None
-            }
-            .andThen { case _ => sessionStoreService.remove() }
-            .map { continueUrlOpt =>
-              val continueUrl = continueUrlOpt.map(_.url).getOrElse(appConfig.agentServicesAccountUrl)
-              val isUrlToASAccount = continueUrlOpt.isEmpty
-              Ok(html.subscription_complete(continueUrl, isUrlToASAccount, prettify(Arn(arn))))
-            }
+          for {
+            continueUrlOpt           <- sessionStoreService.fetchContinueUrl.recover(recoverSessionStoreWithNone)
+            wasEligibleForMappingOpt <- sessionStoreService.fetchMappingEligible.recover(recoverSessionStoreWithNone)
+            _                        <- sessionStoreService.remove()
+          } yield {
+            val continueUrl = continueUrlOpt.map(_.url).getOrElse(appConfig.agentServicesAccountUrl)
+            val isUrlToASAccount = continueUrlOpt.isEmpty
+            val wasEligibleForMapping = wasEligibleForMappingOpt.contains(true)
+            val prettifiedArn = prettify(Arn(arn))
+            Ok(html.subscription_complete(continueUrl, isUrlToASAccount, wasEligibleForMapping, prettifiedArn))
+              .removingFromSession("arn")
+          }
         case _ =>
           Future.successful(sessionMissingRedirect())
       }
     }
   }
+
+  private def linkAccountResponse[A](
```

Comment:
what happens if the mapping was requested, but failed? subscription would be complete at that point, leaving user confused. Just something to consider. 

also a lot of _ , perhaps for comprehension would make this clearer.

Reply:
So I think what you're getting at is that the subscription is made before the '/link-account' page is shown rather than afterwards.
If a mapping call fails after they click Yes to the '/link-account' page, they'd see the usual "There was a problem page, please retry again later". If they retried again later, they'd just be sent straight to the AS homepage and the mapping may or may not have been updated (if not, it'd be cleaned up by the mongo TTL).
If we instead delayed the subscription until after the `/link-account` page, then the mapping calls fail, they'd still see the same "There was a problem page, please retry again later" page and be in the exact same situation.

Reply:
lets move the /link-account page before subscription happens. So if they ever drop out at that page, their account is unaffected. They must complete the journey for any changes to take effect to their account.
so, they must HIT the confirm details button.

Reply:
Will raise a tech debt ticket for this so we don't forget and can discuss it

Reply:
Captured in APB-2891